### PR TITLE
Simplify cluster instantiation in tests

### LINKED ISSUE #406: Simplify Kubernetes cluster instantiation in tests
Currently, the test suite utilizes the Golang packages for K3d and KinD to create clusters internally. This is causing issues as I attempt to write a new E2E test that tests functionality with non-standard `KUBECONFIG` environment variables, since it does a complicated series of checks, validations, and logic. It currently does this:

1. Determine if a cluster is already running, and if it is, use it instead of creating one
2. If a cluster isn't running, create one, using a hard-coded `KUBECONFIG` file (`~/.kube/config`)

I believe it would be far simpler and more scalable to abstract that out into the environment. I propose:

1. Expect (and validate) a cluster to be running already, unless explicitly told to instantiate the built-in K3s cluster

It would add a bit more code to the GitHub Action yaml scripts, since the Action would now need to handle creating a cluster and destroying it after the tests finish, but would make the Golang code in the `test/e2e` folder much simpler and easier to maintain.

### PR DESCRIPTION
**Pull request by RothAndrew**
## Description

Simplifies the handling of kubernetes clusters from this logic:

1. Determine if a cluster is already running, and if it is, use it instead of creating one
1. If a cluster isn't running, create one using the Golang libraries for Kind and/or K3d, using a hard-coded `KUBECONFIG` file (`~/.kube/config`)
    - If TESTDISTRO is not set, run the tests twice, once on KinD, and again on K3d
    - If TESTDISTRO is set to either `kind` or `k3d` only do that one
1. Run the tests
1. Destroy the cluster if it was created and delete the `~/.kube/config` file 

To this:

1. Expect TESTDISTRO env var to be set, with value one of `provided`, `kind`, `k3d`, or `k3s` and error out if it isn't
1. If TESTDISTRO is 
    - `kind` or `k3d` -- create that cluster using the associated golang library, using a temporary file for the KUBECONFIG.
    - `provided` -- Use the provided cluster, using the existing KUBECONFIG
    - `k3s` -- Use `zarf init --components k3s` to create the cluster and `zarf destroy` to delete it (done as part of each test so that each test gets a fresh cluster)
1. Run the tests
1. If not `provided` or `k3s`, destroy the cluster and temporary KUBECONFIG file

The pipeline will provide the KinD cluster, and will expect the tests to create the K3d and K3s clusters

## Related Issue

Fixes #406
Related to #373 (except for needing a new e2e test)

## Type of change

- [X] Refactor (non-breaking change which preserves functionality and increases maintainability)

## Checklist before merging

- [x] Tests have been added/updated as necessary (add the `needs-tests` label)
- [x] Documentation has been updated as necessary (add the `needs-docs` label)
- [x] An ADR has been written as necessary (add the `needs-adr` label) [ [1](https://github.com/joelparkerhenderson/architecture-decision-record) [2](https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions) [3](https://adr.github.io/) ]
- [x] (Optional) Changes have been linted locally with [golangci-lint](https://github.com/golangci/golangci-lint). (NOTE: We haven't turned on lint checks in the pipeline yet so linting may be hard if it shows a lot of lint errors in places that weren't touched by changes. Thus, linting is optional right now.)

### CODE REVIEW

**Comment by RothAndrew** [comment]
I don't think an ADR is needed here, but happy to write one if the team feels it is needed.

Still need to update docs, will do that next after I ensure this is working how I expect it to.

Since the whole thing is about tests, I'll go ahead and pull that label

**Comment by RothAndrew** [comment]
Updated description based on offline conversation with @YrrepNoj

**Comment by RothAndrew** [comment]
@YrrepNoj ready for you to look at this based on what we talked about. Still need to update docs but I think it's mostly there.

**Comment by RothAndrew** [comment]
Ready for review and merge

**Review by YrrepNoj — APPROVED** [review]
Worked together in person to address comments.

**Comment by jeff-mccoy** [comment]
Reading back through all this rn after getting back from spring break, give me a few to verify too before merge please. Also note @RothAndrew you'll probably want to squash your commits as you have an unsigned one blocking merge.

File: .github/workflows/test-kind.yml
Comment (jeff-mccoy) [inline .github/workflows/test-kind.yml:23]: Is there a particular reason to have the test suite create a k3d cluster vs do what we are doing for kind?  I have no preference, just the inconsistency is confusing.

File: Makefile
Comment (jeff-mccoy) [inline Makefile:90]: Really like this pattern of explicitly defining what/how you are testing.

File: cli/internal/k8s/common.go
Comment (jeff-mccoy) [inline cli/internal/k8s/common.go:155]: This is a pretty critical change (I think for the better) to the core of how we talk to the control plane that is well outside the scope of a PR for simplifying tests.  Would like to see more code comments here as to what each step is doing and a comment in the PR on why this change was needed here vs it's own PR as changing core code behavior for a tests is usually and anti-pattern.

File: .github/workflows/test-k3s.yml
Comment (RothAndrew) [reply .github/workflows/test-k3s.yml:17]: IDK if it's necessary to create a convention, unless we are able to automate checks for it.

I added the quotes on a whim, since it felt like the name stood out better in the IDE since they turn green when they are wrapped in quotes. I'm perfectly fine with pulling out the quotes, or adding quotes to the other yaml files that i missed.

File: .github/workflows/test-kind.yml
Comment (RothAndrew) [reply .github/workflows/test-kind.yml:23]: We wanted to make sure the test pipeline tested `TESTDISTRO=provided`. That's the only reason one creates the cluster and one expects one to exist.

File: test/e2e/README.md
Comment (jeff-mccoy) [inline test/e2e/README.md:16]: This might be a good new issue to expose the distro detect stuff on the CLI.  It would be really clean for a `zarf tools what-distro` or something command to tell the user about their kubecontext + detected distro.

File: cli/internal/k8s/common.go
Comment (RothAndrew) [reply cli/internal/k8s/common.go:155]: I can pull it back out into its own PR. It was originally its own PR, but it got incorporated here later.

**Comment by RothAndrew** [comment]
> Also note @RothAndrew you'll probably want to squash your commits as you have an unsigned one blocking merge.

The commit is signed, but for some reason GitHub is not seeing it as verified, even though the same GPG key was used for it as all the other commits. I'm having trouble figuring out how to squash/force-push to fix since there's a Merge Commit in the middle.

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [inline test/e2e/clusters/clusters.go:23]: we already have constants used in the codebase to designate distros that look like they are compatible with this code, [distro.go](https://github.com/defenseunicorns/zarf/blob/master/cli/internal/k8s/distro.go#L10).  Seems like it would be cleaner to use that vs buildout an iota enum + map.

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [inline test/e2e/clusters/clusters.go:121]: Why are we implementing custom logic to do what [common.go](https://github.com/defenseunicorns/zarf/blob/master/cli/internal/k8s/common.go#L104) already does?

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [inline test/e2e/clusters/clusters.go:137]: Concerned about these functions being similar + similar name to the k8s package, is there a reason they had to be custom logic here vs use the core code?  We don't want custom connect logic for tests because then we're cheating the E2E tests.

File: test/e2e/clusters/clusters.go
Comment (YrrepNoj) [reply test/e2e/clusters/clusters.go:121]: This was moved from a different an existing function in the test code that existed before the new cli-supported common.go function. You're right that we should be using that instead now that it exists though.

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [inline test/e2e/clusters/clusters.go:165]: Little confused, I thought that Kind was now handled outside of tests as the docs stated above?

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [inline test/e2e/clusters/clusters.go:238]: Same comments as before, we already have wait logic to wait for a healthy cluster we should be testing, not implementing custom logic to bypass it for tests.

File: test/e2e/clusters/clusters.go
Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:23]: Different need. Here we are getting user input about what cluster they want. There we are determining what cluster Zarf is being run against.

File: test/e2e/main_test.go
Comment (jeff-mccoy) [inline test/e2e/main_test.go:36]: lol like the function name 🤣

File: test/e2e/clusters/clusters.go
Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:121]: The custom logic was already there before this PR. I can look at pulling it out.

**Review by jeff-mccoy — CHANGES_REQUESTED** [review]
Overall I think this will be an improvement to the testing flow, but some custom logic may need to use the main codebase instead and we should be careful to make sure we're not doing something different than how a normal user would use Zarf as that's the point of E2E is to be as real as possible.  Thanks for working this.

File: test/e2e/clusters/clusters.go
Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:137]: This doesn't cheat the E2E tests, they still use the golang application code to do its thing, through the Zarf binary that gets used.

This gets used in the TestMain function to set up the cluster.

File: test/e2e/clusters/clusters.go
Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:165]: If you use `TESTDISTRO=kind` a kind cluster will be created

File: cli/internal/k8s/common.go
Comment (jeff-mccoy) [reply cli/internal/k8s/common.go:155]: I think it can be fine here is you can explain in this PR what's going on.  Also some code comments per line would be nice here because it's k8s magic block box right now.

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [reply test/e2e/clusters/clusters.go:23]: Different usage of the same data in my mind.  One is saying "this is what this cluster is" the other is saying "this is the cluster I want".  But it's still the same data in the end.  Just don't see the need to add logic with same values elsewhere--but I won't lose sleep over this going into the merge.

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [reply test/e2e/clusters/clusters.go:121]: Yeah let's use the k8s package for this, thx

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [reply test/e2e/clusters/clusters.go:165]: copy must have misunderstood that part from the docs, makes sense though.

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [reply test/e2e/clusters/clusters.go:137]: Copy so I guess a better question is how does this differ from the same functions in the k8s package?

File: test/e2e/clusters/clusters.go
Comment (jeff-mccoy) [reply test/e2e/clusters/clusters.go:238]: Okay then put another way, we should use the existing logic in the codebase to wait for a cluster, I think that's already been agreed in a prior comment.

File: test/e2e/clusters/clusters.go
Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:23]: I'd like to leave it for now. It's already gone through a couple of iterations. I'll make an issue in the backlog to revisit it.

File: test/e2e/README.md
Comment (RothAndrew) [reply test/e2e/README.md:16]: Created #412

File: test/e2e/clusters/clusters.go
Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:23]: Created #411

File: test/e2e/clusters/clusters.go
Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:121]: Do we want to move `k8s` out of the `internal` folder so we can do this? That's a bigger decision/change than I'm willing to take on in this PR.

Comment (jeff-mccoy) [reply test/e2e/clusters/clusters.go:121]: It's all one go.mod, would we need to move it out of internal?  I think _we should move it_, but agree that's out of scope for this PR.

Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:121]: image

Might just be an IDE thing, but still not something I want to do even if it is technically possible.

Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:121]: I suggest we stick with what's there for now and updated it as part of #411

Comment (RothAndrew) [reply test/e2e/clusters/clusters.go:121]: @jeff-mccoy are you good with that?

**Review by jeff-mccoy — APPROVED** [review]

**Pull request merged**
